### ISSUE TITLE
regsearch() not working for some registries due to dependence on hard outputLimit

### ISSUE THREAD

**Issue by tomdonaldson** [astropy/pyvo#373]
`regsearch()` works with the default registry (http://reg.g-vo.org/tap), but fails with the MAST registry (http://vao.stsci.edu/RegTAP/TapService.aspx) and the EURO VO (http://registry.euro-vo.org/regtap/tap).  

One of those reasons for the failure is that the `regsearch()` query depends on the service's `capabilities` endpoint defining an `<outputLimit><hard>` value.  The failure is an ungraceful traceback with the message `pyvo.dal.exceptions.DALServiceError: Hard limit not exposed by the service`.  Since that hard limit is not a required part of `capabilities`, it should not be required by the registry search.  

In a `python` environment including `pyvo`, the following `bash` script reproduces the problem:
```
export IVOA_REGISTRY=http://vao.stsci.edu/RegTAP/TapService.aspx

cat << EOF | python

import pyvo
results = pyvo.regsearch(servicetype='scs')
print(results)
EOF
```

**Comment by bsipocz** [comment]
I wonder whether it's worth having at least one test that uses one of the non-default registries?

**Comment by msdemlei** [comment]
On Mon, Oct 10, 2022 at 07:56:14PM -0700, Tom wrote:
> One of those reasons for the failure is that the `regsearch()`
> query depends on the service's `capabilities` endpoint defining an
> `<outputLimit><hard>` value.  The failure is an ungraceful
> traceback with the message `pyvo.dal.exceptions.DALServiceError:
> Hard limit not exposed by the service`.  Since that hard limit is
> not a required part of `capabilities`, it should not be required by
> the registry search.  

Oh dang.  I suspect that's been broken for a long time, as I don't
remember having introduced the maxrec setting.

First, the background: many TAP servers set rather strict default
MAXREC-s; on mine, this used to be 2000 at about the time the basic
RegTAP interface was written (or so I seem to remember). With limits
like these, there were realistic registry queries that were
truncated.

Thinking of a smart way to handle these truncations in regsearch
seemed inconvenient at the time, and so it seemed like a good idea to
just pass in the hard limit as MAXREC, in particular because the
queries regsearch executes won't return catastrophically large result
sets.

By now, even my service has a default MAXREC of 20000, and it's hard
to imagine that a truncation at that level would hurt a useful
registry query.  Hence, I'd not object to just patching out 
  
  maxrec=service.hardlimit

in regtap.py, line 149.

*However*, I consider that pattern ("give me as many records as you
can, I know what I'm doing, promised") really useful in other
contexts, and hence I *think* we should fix this on the TAP side.
Since the various query constructors all use None as a null value for
MAXREC, I'd say at least the hardlimit property of tap.TAPService
should just return None if it can't work out a hard limit on the
server rather than raise an error.

I *think* I'd even do the same thing for the maxrec property itself.
The use cases for it are less clear to me, but just guessing I'd say
a sentinel value None is just about as good as raising an error, and
consistency between hardlimit and maxrec would be a good thing.

Of course, both changes *will* change exposed behaviour, which
patching out maxrec won't (except hopefully fixing this bug).  On the
other hand, I'd argue that it's saner and more useful behaviour and
that it's unlikely that any actual code will rely on these
exceptions.

So... I'm happy to write a PR for either solution (or both).  But I'd
*really* appreciate guidance: Which of

(1) patch out maxrec passing
(2) returning None from hardlimit and/or maxrec

would you prefer?

And yes, we should probably have a test against one of the other
RegTAP endpoints.  Or perhaps even against everything in
pyvo.regsearch(datamodel='regtap')?

**Comment by tomdonaldson** [comment]
> I wonder whether it's worth having at least one test that uses one of the non-default registries?

Yes, I think we should do that, but it will have to be a work in progress for now.  After removing the hard limit requirement, both the MAST and EURO VO registries fail (for separate reasons) on the actual RegTAP query, so it would be hard to get the test to pass.  Addressing those failures may require changes on the service side.

**Comment by tomdonaldson** [comment]
> So... I'm happy to write a PR for either solution (or both). But I'd *really* appreciate guidance: Which of 

> 1. patch out maxrec passing 
> 1. returning None from hardlimit and/or maxrec

@msdemlei you raise some good points there.  I think both is probably the best ultimate answer, but I've made PR #374 for option 1 for now.  

In addition to the behavior changes, option 2 raises a question I'd like to look at more carefully about whether or how to inform the user that they are requesting (or may have hit) a limit.  Though the limits can be hit now anyway, so maybe that's a separate question.

**Comment by andamian** [comment]
`RegistryResults` will warn on overflow which I think it's the best to go about this.  

I think `search` should expose `MAXREC` as an argument to allow users to sample a large result set without necessary downloading it. The `maxrec` specified by the user could then be compared with `service.hardlimit` before sending it to the service.

**Comment by tomdonaldson** [comment]
This is fixed with #375.
